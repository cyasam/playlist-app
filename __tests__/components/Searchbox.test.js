import React from 'react'
import { shallow } from 'enzyme'
import { Searchbox } from '../../src/scripts/components/Searchbox'

describe('Searchbox Component', () => {
  const setup = (propsOverride) => {
    const props = {
      fetchSearch: jest.fn(),
      history: { 
        location: { pathname: '/' },
        push: jest.fn()
      }, ...propsOverride
    }
    const wrapper = shallow(<Searchbox {...props} />)

    return {
      wrapper,
      props
    }
  }

  it('renders properly', () => {
    const { wrapper } = setup()

    expect(wrapper).toMatchSnapshot()
  })

  describe('Form elements', () => {
    it('has form element', () => {
      const { wrapper } = setup()
      expect(wrapper.find('form').exists()).toBe(true)
    })

    it('has input text element', () => {
      const { wrapper } = setup()
      expect(wrapper.find('input[type="text"]').exists()).toBe(true)
    })

    it('has submit button', () => {
      const { wrapper } = setup()
      expect(wrapper.find('button[type="submit"]').exists()).toBe(true)
    })
  })

  describe('Form input element', () => {
    it('loads initial state', () => {
      const { wrapper } = setup()
      expect(wrapper.state().input).toEqual('')
    })
    
    it('loads initial state if page is search', () => {
      const { wrapper, props } = setup({
        history: { 
          location: {
            pathname: '/search',
            search: `?query=abc`
          }
        }
      })
      expect(wrapper.state().input).toEqual('abc')
    })

    it('saves input value into state when typing', () => {
      const { wrapper } = setup()
      wrapper.find('input').simulate('change', {target: { value: 'abc' }})
      expect(wrapper.state().input).toEqual('abc')
    })
  })

  describe('Form element', () => {
    it('fire handleSubmit event when submitting', () => {
      const { wrapper, props } = setup()
      wrapper.find('form').simulate('submit', { preventDefault: jest.fn()})
      expect(props.history.push).toHaveBeenCalled()
      expect(props.fetchSearch).toHaveBeenCalledWith(wrapper.state().input)
    })
  })
})