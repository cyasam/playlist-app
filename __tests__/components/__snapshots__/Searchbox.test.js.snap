// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Searchbox Component renders properly 1`] = `
<form
  className="search-form mr-auto"
  onSubmit={[Function]}
>
  <input
    className="form-control mr-sm-2"
    onChange={[Function]}
    placeholder="Search videos on youtube"
    type="text"
    value=""
  />
  <button
    className="btn btn-secondary my-2 my-sm-0"
    type="submit"
  >
    <img
      alt="Search"
      height="16"
      src={null}
      width="16"
    />
  </button>
</form>
`;
